오픈마켓 [Step1] - 예스코치

YesCoach opened the pull request:
안녕하세요 찌루루! @jae57 
Step1 진행하고 이렇게 PR을 보내게 되었습니다!!
앞서 진행했던 프로젝트지만 URLSession을 통한 네트워크 통신, 유닛 테스트가 어려워서 다시 공부하면서 진행하고 있습니다🥲
이전 코드와 구조는 대부분 비슷할 것 같아요ㅠㅠ😅 미숙한 부분이나 개선할 사항이 있다면 폭풍 지적 부탁드립니다ㅎㅎ

# 구현한 내용
### Model
1. 서버로부터 받은 데이터를 파싱하기 위해 decodable한 `items`와 `item`을 구현했습니다.
2. 서버로 데이터를 보낼때를 위해서 encodable한 `DELETEItem`과 멀티파트폼에 대응시킬 `POSTItem`, `PATCHItem`을 구현했습니다.
3. 이미지를 서버로 보낼때 Request의 body에 적절한 양식으로 넣어주기 위해 `Media` 타입을 구현했습니다.
### NetWork
1. 서버와 통신을 담당하는 `NetworkManager`를 구현했습니다.
네트워크와 무관한 테스트를 진행하기 위해서 URLSession에 의존성 주입을 추가했습니다.
2. 서버와 통신하는 메소드를 하나만 구현해서 쓰고 싶어서 URL과 Request를 구성하는 API 타입을 매개변수로 받게 했습니다.
Request의 body가 필요한 경우와 필요치 않은 경우로 나뉘기 때문에 하나의 프로토콜을 만들고, 이를 채택하는 두가지 경우의 프로토콜로 세분화 했습니다.
```swift
protocol Requestable {}
protocol RequestableWithoutBody: Requestable {}
protocol RequestableWithBody: Requestable {}
```
3. URL과 Request, Body를 구성하는 로직은 하단에 extension으로 분리 구현했습니다.
4. `Data` 타입에서 Data를 append 할 때 문자열을 자동으로 형변환해서 추가할 수 있도록 확장 구현했습니다.
5. 각각의 기능에 해당하는 `RequestAPI` 타입과  `NetworkError`타입, RequestAPI 구성에 사용하는 Enum을 한곳으로 모았습니다.
### Controller
1. 이름을 통해 `NSDataAsset`을 반환해주는 기능을 포함한 `CustomJSONDecoder`를 구현했습니다. 목데이터를 받아올때 쓰고 있습니다.
2. 뷰컨트롤러에 네트워크랑 통신하는 코드들을 작성하고 주석처리 했었는데, 사용하지 않는 주석은 남기지 말라는 내용이 생각나서 PR 작성하다가 지웠습니다... PR 쓰면서도 여러번 커밋하게 됬네요,,
### Test
1. 네트워크와 무관한 테스트를 위해 `MockURLSessionDataTask` 및 `MockURLSession`를 구현했습니다.
2. 테스트에서는 올바르게 데이터를 파싱할 수 있는지, Response에 따라 통신의 결과가 알맞게 나오는지(상태코드를 기준으로 진행했습니다), 마지막으로 실제 서버로부터 데이터가 잘 받아지는지(통신이 이루어지는지)를 테스트 했습니다.
### Open Source Library
이번에 활동학습때 의존성 관리도구를 배웠고, 프로젝트에 `SwiftLint`를 적용해보라고 하셔서 적용해보았습니다! CocoaPods을 사용했습니다😅

# 고민되는 부분
1. Test 관련
테스트는 항상 어렵네요.. 특히 네트워크 관련 테스트에서 URL를 구성하는 부분이나 Request를 구성하는 부분도 개별적으로 테스트 해보고 싶었는데요, NetworkManager에서 private으로 은닉화를 시켜줬기 때문에 결국 `commuteWithAPI` 메서드만 통째로 테스트하게 되었습니다.
저번 프로젝트에서는의도치 않게 성공한 무의미한 테스트가 있었기 때문에 지금의 테스트 코드가 구현된 기능등을 제대로 검증할 수 있는지 궁금합니다!! 또한 추가로 테스트 해볼 수 있는 내용이 있을지 궁금합니다.
2. 프로젝트 파일명 및 그룹 분류 관련
다른사람이 보고 바로 이해가 된다면 올바른 네이밍이겠죠..? 특히 `Model` 과 `Network`내부의  네이밍이 적절한지 궁금합니다!!
또한 이전 프로젝트에서는 구조체, 클래스 별로 파일을 만들었었는데 프로젝트가 많이 커져서(?) 이번에는 같은 타입이면 한 파일에 모아보았습니다 (특히 `RequestAPI`)
이러한 네이밍과 분류, 위치가 적절한지 궁금합니다.
3. 에러처리
프로젝트를 진행하면서 에러처리는 대부분 `try?`를 쓴 다음에, 해당 메소드가 nil이 나올때 처리하는 식으로 진행했습니다.
이상적인 방향은 do catch문을 통해 특정 에러에 대한 catch를 진행하는게 맞겠죠?? 에러 타입을 생성하고 던지는 것 까지는 하는데 항상 에러처리가 제대로 안되고 뭉뚱그려 하는 것 같아서 고민입니다. 어떻게 에러처리를 하는게 좋을까요?
---
PR 적으면서 코드 수정도 여러번 한 것 같아요🥲 첫 피알 잘부탁드립니다!!😀😀

YesCoach committed:
feat: JSON 데이터를 파싱할 모델 타입 구현

codable을 채택한 타입으로 Item과 Items 구현
MVC 역할에 따른 파일 분리

YesCoach committed:
feat: Media 타입 구현

multi-part formdata 구성에 필요한 미디어 타입 구현

YesCoach committed:
feat: reqeust body를 구성할 세가지 타입 구현
POSTItem, PATCHItem, DELETEItem 타입 구현

YesCoach committed:
feat: NetworkManager 타입 생성 및 메소드 1차 구현

YesCoach committed:
feat: NetworkManager 통신 담당 메소드 구현

YesCoach committed:
feat: NetworkManager 서버 통신 기능 구현
createRequest 메서드 기능 분리
URL, Request 및 RequestBody 생성하는 메서드 구현

YesCoach committed:
refactor: createRequeestBody 로직 수정
multipart 형식에 맞게 requestBody 맨 마지막 열에 코드 추가

YesCoach committed:
refactor: NetworkManager, Model, RequestAPI 리팩토링
RequestAPI에서 parameters 프로퍼티의 타입 변경(옵셔널 해제).
이에 따라 NetworkManager에서 requestBody 로직 개선(parameters옵셔널바인딩제거).
ViewController에서 서버 통신 사용코드 추가(주석처리)

YesCoach committed:
feat: Test를 위한 Mock 타입 및 Mock 데이터 생성

YesCoach committed:
test: Unit Test 작성
Protocol 타입 폴더로 분리
네트워크 무관 테스트 진행을 위해 MockURLSession 수정
Parsing과 Network Commute 메소드 관련 네트워크 무관 테스트 진행

YesCoach committed:
feat: SwiftLint 라이브러리 추가
의존성 관리도구 실습차 프로젝트에 SwiftLint 라이브러리 추가
SwiftLint의 컨벤션에 맞게 코드 스타일링

YesCoach committed:
chore: .gitinit에 Pods 추가하여 Pod 파일 업로드 방지

YesCoach committed:
refactor: VC에 모든 요청 메소드 사용 코드 추가(주석)

YesCoach committed:
refactor: 변수, 매개변수명 수정 및 불필요한 주석 삭제
Coding Convention에 맞춰 변수 및 매개변수명 수정(
ViewControlle의 불필요한 주석들 제거

YesCoach committed:
chore: NetworkManager -> Network 그룹명 수정

YesCoach committed:
test: 실제 네트워크와 테스트하는 코드 추가

YesCoach committed:
refactor: 사용하지 않는 주석 제거

jae57 commented on OpenMarket/OpenMarket/Controller/CustomJSONDecoder.swift:
```diff
@@ -0,0 +1,40 @@
+}
+
+class CustomJSONDecoder: JSONDecoder {
+    static func receiveDataAsset(assetName: String) throws -> NSDataAsset {
```
receive는 데이터를 주는 객체가 있고, 그 데이터를 받는 느낌이 나네요! 
좀 더 적절한 네이밍을 고민해보면 좋을 거 같아요!

YesCoach replied:
Assets에 있는 데이터를 이름을 통해 가져오는 로직인데 `fetchFromAssets`는 어떨까요??

jae57 commented on OpenMarket/OpenMarket/Model/PATCHItem.swift:
```diff
@@ -0,0 +1,51 @@
+    private let price: Int?
+    private let currency: String?
+    private let stock: Int?
+    private let discounted_price: Int?
```
snake case 로 쓰신 이유는 모엇인가요?

YesCoach replied:
서버 API에 명세되어있는 변수명으로 구성하다보니까 snake case로 쓴 것 같아요!!
하지만 생각해보니 request body를 구성할때 저 이름으로 주고, 프로퍼티 자체는 snake case를 쓰지 않아도 될 것 같네요!!
`discountedPrice` 로 변경하고, 하단 parameter() 메소드에서 `parameter["discounted_price"] = discountedPrice` 를 수행하는 쪽으로 수정했습니다🙂

jae57 commented on OpenMarket/OpenMarket/Network/APIEnum.swift:
```diff
@@ -0,0 +1,75 @@
+        case .patch(let id):
+            return Self.baseURL + "/item/\(id)"
+        case .delete(let id):
+            return Self.baseURL + "/item/\(id)"
```
세가지 case 의 동작이 동일하니 하나로 묶어도 되겠네요!

YesCoach replied:
앗 맞습니다!! 완전 동일하네요ㅠㅠ

jae57 commented on OpenMarket/OpenMarket/Network/NetworkManager.swift:
```diff
@@ -0,0 +1,109 @@
+
+class NetworkManager {
+    private let session: URLSessionProtocol
+    lazy var boundary = generateBoundary()
```
computedProperty 로 만들면 아래까지 스크롤을 내리지 않고도 바로 boundary 의 형태를 알 수 있을거 같아요!

YesCoach replied:
재사용을 할 필요 없다면 연산프로퍼티로 쓰는 방법도 좋을 것 같다는 생각이 들었습니다..!

연산프로퍼티로 적용해봤는데 이슈가 생겼습니다ㅜㅜ
multi-part 폼을 구성할때 쓰는 저 boundary는 (하나의 body에서는) 항상 동일한 문자열이여야 하는데, 연산프로퍼티로 구현할 경우 해당 변수를 사용할때마다 새로운 문자열이 만들어졌고, 서버에서 실패한 응답코드를 보내왔습니다.
```swift
return "Boundary-\(UUID().uuidString)"
```
정확히 이해한 것인지는 모르겠지만, 
_연산 프로퍼티는 호출할 때마다 연산을 실행하므로,  `UUID().uuidString`이 매번 새로운 무작위 문자열을 생성하게 되어 서버에서 에러를 보내왔다_
이렇게 이해하면 될까요??

jae57 replied:
네!! 이해하신게 맞습니다! 👍     
매번 새로운 패턴의 문자열이 생성되는 거라 그렇게 구현하셨던 거군요!! 
이부분은 제가 체크를 못했네요;; 🙇‍♂️

YesCoach replied:
감사합니다!! 연산프로퍼티 사용을 거의 안해왔었는데 이번 기회에 확실히 이해하고 넘어갈 수 있었습니다!!

jae57 commented on OpenMarket/OpenMarket/Network/NetworkManager.swift:
```diff
@@ -0,0 +1,109 @@
+            }
+            guard let response = response as? HTTPURLResponse, (200...299).contains(response.statusCode) else {
+                return completion(.failure(NetworkError.invalidResponse))
+            }
```
response는 guard 문 이후로 쓰이는 곳이 없으니 if 문으로 해당 케이스를 쳐내는 것도 좋을거같네요!

YesCoach replied:
if문과 guard문 차이를 생각못했던것 같아요! if문으로 수정하겠습니다.
이런 부분은 코드가 길어졌을때를 위해서라도 꼭 지켜주는게 좋은걸까요??
(guard 문으로 작성시 해당 변수가 밑에서 또 사용된다는 의미를 내포하므로..?)

jae57 replied:
이부분은 아무래도 상황에 따라 다르고, 보는 사람에 따라 다른 부분이라    
꼭 지키시라고 말씀드린건 아닙니닷 😋   
if 문으로 작성했을 때, guard 문으로 작성했을 때 각각을 생각해보시고   
둘 중 어떤게 가독성이나 코드복잡도가 낮아질까를 생각해보셨으면 해서 코멘트 남겼습니다!

jae57 commented on OpenMarket/OpenMarket/Network/NetworkManager.swift:
```diff
@@ -0,0 +1,109 @@
+extension NetworkManager {
+    private func createURL(from api: Requestable) -> URL? {
+        let url = URL(string: api.url.description)
+        return url
```
변수를 만들필요 없이 바로 return 해줘도 되겠네요!
api 의 url 을 구성하는 부분이 이미 enum 내에 있고 해당 메서드에서는 가져오는 것 뿐이니 따로 함수로 떼어내지 않아도 될거같아요!

YesCoach replied:
수정했습니다!! 기능별로 분리해야겠다고 생각했었는데 url 문자열을 가져와서 URL로 만들어주는 기능은 굳이 함수로 분리할 필요까지는 없을것 같습니다! (써놓고 보니 한줄이네요...)

jae57 commented on OpenMarket/OpenMarket/Network/NetworkManager.swift:
```diff
@@ -0,0 +1,109 @@
+        return url
+    }
+
+    private func createRequest(api: Requestable) throws -> URLRequest {
```
Request 를 만드는 메서드를 하나로 만들어서 코드 중복은 많이 줄었지만 그만큼 다양한 api 들에 대응하려다 보니 
if 문이 늘어났고, 앞으로 api 가 추가된다면 더 늘어날 수 있을거 같아요!
해당 메서드가 현재 어떤 역할들을 가지고 있는지 생각해보시고, 쪼개볼 수 있는 부분은 쪼개 보면 좋을 거 같아요!
이를테면, body 가 필요한 request를 만드는 메서드와  body가 없어도 되는 request를 만드는 메서드로 나눠 본다던지요!

YesCoach replied:
음... if 등으로 분기를 하는것이 결국 SRP 같은 부분을 지키지 못하게 되는 코드로 이어진다는 말도 들어본 것 같아요
어떤 개발자분들은 분기를 피하기 위해 프로토콜을 사용한다고 하셨는데 그부분을 고민하다가 시간을 많이 뺐겼습니다😭
리팩토링한 코드는 어떻게 보시는지 궁금합니다! 그리고 프로토콜로 분기한다는게 어떤 방향인지 궁금합니다ㅜㅜ 어떤 방법이 있을까요?

jae57 replied:
https://soojin.ro/blog/solid-principles-in-swift 의 OCP 설명부분의 enum 을 이용한 방법과
https://medium.com/swift-fox/refactoring-replace-enum-with-polymorphism-c4803baeba07 의 protocol 을 이용한 방법 참고해보시면 좋겠습니다!

jae57 commented on OpenMarket/OpenMarket/Network/RequestAPI.swift:
```diff
@@ -0,0 +1,67 @@
+    }
+}
+
+struct DeleteItemAPI: Requestable {
```
deleteItemAPI 는 `RequestableWithBody` 가 아닌 `Requestable`을 채택한 이유가 있을까요?

YesCoach replied:
`deleteItemAPI`는 body가 JSON 형태인데, `RequestableWithBody`는  멀티파트폼에 대응하고 있어요! 그래서 `Requestable`을 채택하도록 하고 JSON에 대응하도록 추가 구현을 해줬는데요, 생각해보니 이것도 분리를 해주는게 좋을 것 같습니다. DeleteItemAPI는 body가 있는데 이렇게 해버리면 코드를 이해하는데 힘들 것 같아요.
프로토콜을 추가로 분리해보겠습니다!!

jae57 commented on OpenMarket/OpenMarket/Network/RequestAPI.swift:
```diff
@@ -0,0 +1,67 @@
+    var url: APIURL
+    var method: APIMethod = .delete
+    var contentType: ContentType = .json
+    var deleteItem: DELETEItem
```
네이밍 상으로는 뭔가 지워져야할 정보가 담겨있을 거 같은데 password 가 들어있네요?
요청하는 사람을 인증할 수 있는 정보니 authInfo 와 같은 네이밍은 어떨까요?

YesCoach replied:
어... 말씀 듣고보니 그렇네요?!? 결국 필요한건 DELETEItem의 password 값이니까, deleteItem을 `authoInfo`로 변경하면 좋을 것 같아요!! 그럼 의미 전달이 제대로 될 것 같아요.

YesCoach replied:
`RequestableWithBody`를 JSON과 Multipart에 대응하는 프로토콜로 세분화하면서 Multipart에서는 `parameters`이라는 네이밍을 쓴 것을 보고, JSON에서는 `json`으로 하는게 좋을까 싶어서 `json`으로 변경해보았습니다!

jae57 replied:
> 어... 말씀 듣고보니 그렇네요?!? 결국 필요한건 DELETEItem의 password 값이니까, deleteItem을 `authoInfo`로 변경하면 좋을 것 같아요!! 그럼 의미 전달이 제대로 될 것 같아요.

요부분은 아직 적용이 안된걸까요?

jae57 commented on OpenMarket/OpenMarketTests/Mock/MockURLSession.swift:
```diff
@@ -0,0 +1,34 @@
+
+    func dataTask(with request: URLRequest, completionHandler: @escaping (Data?, URLResponse?, Error?) -> Void) -> URLSessionDataTask {
+        let successResponse = HTTPURLResponse(url: url, statusCode: 200, httpVersion: "2", headerFields: nil)
+        let failureResponse = HTTPURLResponse(url: url, statusCode: 402, httpVersion: "2", headerFields: nil)
```
매번 함수 호출 시 마다 생성하기보다 전역으로 빼서 한번 생성하고 재사용 하거나
아래 if 문 아래에 포함시켜서 필요할 때만 생성되도록 해주면 가독성면에서도 더 좋을거 같습니다!

YesCoach replied:
생각못했는데 둘 중 하나는 생성해도 안쓰이니까요!! if문에 포함시켜 필요할때 생성하는 로직으로 수정하겠습니다✅

jae57 commented on OpenMarket/OpenMarketTests/OpenMarketTests.swift:
```diff
@@ -0,0 +1,124 @@
+@testable import OpenMarket
+
+class OpenMarketTests: XCTestCase {
+    let sut: NetworkManager! = nil
```
선언은 되어있지만 쓰이는 곳은 없네요??

jae57 commented on OpenMarket/OpenMarketTests/OpenMarketTests.swift:
```diff
@@ -0,0 +1,124 @@
+import XCTest
+@testable import OpenMarket
+
+class OpenMarketTests: XCTestCase {
```
네이밍을 좀 더 구체적으로 바꿔보면 어떨까요?

YesCoach replied:
지금까지는 단순히 프로젝트명 뒤에 Tests가 추가된 디폴트네이밍을 사용했는데요 (bankManagerTests, CalculatorTests 등 자동완성된 이름), 이 테스트가 어떤 내용을 테스트하고 있는지를 포함하는 네이밍이 좋을까요???
네이밍 뿐 아니라 테스트하는 내용에 맞게 파일을 분리해보았습니다!
`NetworkManagerTests` `ParsingDataTests`

jae57 commented on OpenMarket/OpenMarketTests/OpenMarketTests.swift:
```diff
@@ -0,0 +1,124 @@
+        XCTAssertEqual(result, expectResult)
+    }
+
+    func testNetworkManager_commuteFromMockNetwork_successs() {
```
테스트 네이밍에 일관된 규칙을 가져가면 좋을 거 같습니다!
위에서는 _ 뒤에 대문자로 시작하는데 여기선 소문자라서요..!ㅎㅎ 단순 오타인가요?

YesCoach replied:
오타... 에요ㅜㅜ

jae57 commented on OpenMarket/OpenMarketTests/OpenMarketTests.swift:
```diff
@@ -0,0 +1,124 @@
+        // give
+        let testTarget = "MockItem"
+
+        guard let json = try? CustomJSONDecoder.receiveDataAsset(assetName: testTarget) else {
+            return XCTFail("json 파일을 찾을 수 없습니다.")
+        }
+        // when
+        guard let decodedData = try? CustomJSONDecoder().decode(Item.self, from: json.data) else {
+            return XCTFail("decoding에 실패했습니다.")
+        }
```
모든 테스트에 해당 부분이 중복되는데 해결해 볼 수 있을까요?
`setUpWithError()` , `setUp()` , `tearDown()`, `tearDownWithError()` 메서드에 대해 알아보고 적용해봅시다!

YesCoach replied:
setUP, tearDown에 대해서, 모든 테스트가 깨끗한 상태(state)로 시작하는지 확인하기 위해서 setUp() 에서 SUT(System Under Test) 객체를 생성하고, tearDown() 에서 해제하는 것이 가장 좋은 방법이라는 것을 인지했습니다!!
근데 에셋을 받아오는 부분이나, 디코딩하는 부분이 매개변수로 인해 조금씩 차이가 있더라구요! 네트워크 통신 관련 테스트에서도 `URLSession`을 쓸 건지, `mockURLSession`을 쓸 건지, `mockURLSession`이라면 성공한 경우를 가정할 건지 실패한 경우를 가정할 건지에 따라 선언이 달라지고 있기에 적용하는데는 조금 무리가 있었습니다ㅜㅜ

그래서 일단은 서버와 통신하는 객체 선언과 decoding을 수행할 객체만 초기설정 및 종료시 메모리 해제 처리 해주었습니다

jae57 replied:
앗.. 그렇네요!! 좋습니다!

jae57 commented on OpenMarket/OpenMarketTests/OpenMarketTests.swift:
```diff
@@ -0,0 +1,124 @@
+        let page = 1
+        let sut = NetworkManager()
+        // when
+        sut.commuteWithAPI(with: GetItemsAPI(page: page)) { result in
```
메서드 명에 withAPI 가 이미있으니 파라미터 네임을 아예 생략해도 충분히 이해할 수 있을 거 같아요!

jae57 commented on OpenMarket/OpenMarketTests/Mock/MockItems.json:
```diff
@@ -0,0 +1,251 @@
+{
```
MockItems.json 과 MockItem.json 이 여기에도 있고, Assets.xcassets 에도 있는데 두 곳에 있는 이유가 따로 있을까요?

YesCoach replied:
음... 사실 잘 모르고 썼어요🥲  목데이터를 프로젝트에 추가하고 싶어서 일단 json 파일을 만들었고, 이걸 다시 Assets에 추가했어요😅  지금 코드에서는 Assets를 참조해서 받아오고 있으니 Mock에 있는 json은 삭제해도 괜찮을까요? 별개로 프로젝트에서 데이터 에셋을 관리하는 방법을 한번 더 공부해보겠습니다...!

jae57 replied:
네! 삭제해도 문제없을 거 같습니다!

jae57 reviewed (commented):
안녕하세요 예스코치 🙋‍♂️
@YesCoach 

테스트 잘 작성해주셔서 전반적으로 코드 이해에 도움이 많이 되었습니다!
고민사항 남겨주신 부분 잘 읽어보았습니다. 
관련하여 제 의견 남겨보았으니 읽어보시고 추상적이거나 이해되지않는 부분은 편하게 DM 주세요

**1. 테스트에 관하여**
private 이라 테스트할 수 없는 부분은 일단 놔두고 할 수 있는 부분에 집중해봅시다!
우선 궁금한 것이 테스트를 아이템 조회, 리스트 조회 API 에 대해서만 짜두셨는데 
나머지 추가, 수정, 삭제 API 에 대한 테스트를 짜보는 건 어떨까요?
이미 짜두신 리스트 조회도 다양한 케이스에 대해 테스트해보면 좋을 거 같습니다!
( 예를 들면, 페이지를 달리해보면서 일관된 개수의 아이템들이 넘어오는지, 마지막 페이지에는 몇개가 오는지 )
에러메시지가 잘 나오는지에 대한 테스트도 추가해보면 좋겠구요!

**2. 프로젝트 파일명 및 그룹 분류에 관하여**
목적에 맞게 잘 그룹화해준 거 같아요! 
Model 과 Network내부의 네이밍에 관해서는 POST, PATCH, DELETE 등이 조금 생소하긴 했지만 문제될 건 없습니다!
RequestAPI 파일에 관련 API 들을 모아둔 것도 찾아보기 편해서 굿좝입니다.
다만, 모아둔 김에 네이밍에 API 가 반복되니 아예 namespace 를 하나 감싸줘보면 어떨까요? 
API.GetItem(id: 1), API.GetItems(page:2) 이런식으로 사용할 수 있게요!
전반적으로 extension 들은 따로 떼어서 아래부분에 모아두셨던데( 예를 들면, APIEnum 파일 ) 
적당히 길지않은 코드라면 함께 모아두면 코드 볼 때 위 아래를 왔다갔다하지 않고도 이해할 수 있으니 편할 거 같습니다! 

**3. 에러처리에 관하여**
우선 do catch문을 통한 처리가 항상 이상적인건 아니라고 생각합니다ㅎㅎ 
뻔하지만 때에 따라 다를 거 같아요 에러를 처리하는 사람 마음이죠!
이번기회에 try? 와 do catch 문은 어떤 차이가 있는지 알아보고 
각각 어느 경우에 활용되는지 코드들을 보면서 감을 익혀보면 좋을 거같아요!
에러를 던지기까지는 했는데 아직 STEP 1 이라서 api 응답결과를 사용하는 곳이 없다보니 던져진 에러를 처리를 제대로 처리하지 않은 느낌이 난거 같습니다! 스텝을 더 진행하면서 고민해나가면 될거같아요!
한가지 더 붙이자면 에러를 항상 던질 필요는 없습니다! 
보통 던지는 경우는 해당 에러를 다른 곳에서 처리해주었으면 할 때 입니다!
예를 들면 이 에러를 사용자에게 보여주고 싶어! 라고 하면 view 를 그리는 곳까지 에러가 전파되어야겠죠!
하지만 그렇지 않고, 요 에러는 발생하면 그냥 내가 디버깅용으로만 보고싶어! 라고 하면 당장 debugPrint로 에러메시지만 찍어두도록 처리해버릴 수도 있습니다!


코드에도 코멘트를 남겼는데
제 의견을 써놓은 거라 동의하시는 부분만 쏙쏙 골라서 적용해주시고  
저와 다른 의견 환영합니다! ㅎㅎ

YesCoach committed:
refactor: 네이밍 수정 및 NetworkManager 함수 리팩토링
CustomJSONDecoder에서 의미를 분명히하기 위해 receiveDataAsset 이름 수정
PATCHItem의 snake case가 사용된 프로퍼티명 수정
APIEnum의 URL에서 switch문 내 동일한 동작을 하는 case들 하나로 통합
NetworkManager의 일부 메소드 제거 혹은 연산 프로퍼티 적용
NetworkManager의 session.dataTask에서 response 처리 로직 수정

YesCoach committed:
refactor: NetworkManager의 request 메소드 로직수정
createRequestBody 메소드 body의 타입에 따른 로직을 별개의 메소드로 분리
NetworkError에 에러 case 추가
RequestableWithBody를 두 개의 프로토콜로 세분화

YesCoach committed:
refactor: RequestAPI내 타입들에 대해 namespace 적용

YesCoach committed:
test: test case 추가 및 test 분리
NetworkManager의 test와 JSONDecoder의 test로 파일 분리
NetworkManager에서 PUSH, PATCH, DELETE 메소드에 대한 테스트 코드 추가

YesCoach committed:
test: NetworkManagerTests에 tearDownWithError 추가

YesCoach committed:
refactor: NetworkManager의 commuteWithAPI 수정
이름이 중복되므로 파라미터 네임 삭제

YesCoach committed:
docs: Mock 폴더 내 json 파일 삭제
Assets에 이미 존재하므로 중복되는 데이터인 Mock 내 json 파일 삭제
Test내 발생한 에러 수정(commuteWithAPI 메소드의 파라미터 관련)

YesCoach committed:
test: GETItems시 몇개의 item이 받아지는지 테스트

YesCoach committed:
test: test 타입 네이밍 변경

jae57 commented on OpenMarket/OpenMarketTests/NetworkManagerTests.swift:
```diff
@@ -0,0 +1,162 @@
+        sut = nil
+    }
+
+    func test_commuteFromMockNetwork_successs() {
```
오타발견이요! sss

YesCoach replied:
으앙...

jae57 commented:
예스코치, 안녕하세요! @YesCoach 
제가 남긴 리뷰들 최대한 반영하려 노력해주시는게 보여 감동입니다🥺

테스트도 목적에 따라 분리하시고, 더 많은 케이스를 추가해주셨네요! 👏
실제로 API를 이용하다보면 더 많은 테스트 케이스들을 생각해볼 수 있을테니
테스트는 다음 스텝들을 진행하면서 더 발전시켜봅시다!

브랜치 문제만 해결되면 바로 머지하겠습니다!
고생하셨습니다! 👍👍
